correct type of boolean operators

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
correct type of boolean operators

tautschnig commented on regression/goto-analyzer/intervals_01/main.c:
```diff
@@ -1,3 +1,4 @@
+#include <assert.h>
```
Ok, but how does this relate to the PR? Adding a body to the commit message might be really helpful to explain this.

kroening replied:
Unrelated -- will drop.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -306,6 +306,8 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
     symbol_exprt bound=to_symbol_expr(expr.op0().op0());
     expr.op0().swap(bound);
+
+    implicit_typecast_bool(expr.op1());
```
Could we have a regression test for this? I suppose that might be fairly easy to construct from one of the existing tests using quantifiers?

kroening replied:
Covered by existing tests; I've added a note about that into the commit message.

tautschnig replied:
Hmm, if the existing tests cover it, why does this change have no impact (such as turning a test from KNOWNBUG to CORE)?

kroening committed:
new simplifier rule '(bvtype)c' --> 'c?1:0' for boolean 'c'

kroening committed:
revive simplifier rule (T)(c?x:y) --> c?(T)x:(T)y

This rule was considered in the past, but removed again as this
simplification may yield a more expensive formula if the complexity of 'T'
is higher than that of the if expression (say 'double' vs.  'int').

Thus, we only enable when x and y are constant.  This enables re-writing,
for instance, (long)c?1:0 to c?1L:0L.

kroening committed:
use simplifier in ai_simplify

smowton commented:
To check: will this now crash with `--no-simplify` because the backend doesn't know how to handle bool-to-int casts?

tautschnig commented:
@kroening I'm not sure whether MC/DC coverage is still handled correctly, these are now the only test failures left.

TGWDB commented:
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.
